### ISSUE TITLE
On Facebook, sharing a link to a Grapher chart with a query string that has encoded spaces is broken

### ISSUE THREAD

**Issue by breck7** [owid/owid-grapher#397]
A user reported that sharing certain links is broken by Facebook. I was able to repro.

Facebook will add a "fbclid" param to the query string. That doesn't break things. The problem is they will also rewrite our query string params, turning "North%20America" into "North+America".

This:
https://ourworldindata.org/grapher/daily-cases-covid-19?country=Africa+Asia+Europe+North%20America+Oceania+South%20America+OWID_WRL

Becomes this:
https://ourworldindata.org/grapher/daily-cases-covid-19?country=Africa+Asia+Europe+North+America+Oceania+South+America+OWID_WRL&fbclid=some_long_random_id

I couldn't find any documentation from FB on this. Not sure if a lot of people are hitting this and whether it's worth changing the way we do query strings which might be tricky.

**Comment by jasoncrawford** [comment]
That is a weird way of doing query strings. We shouldn't be using pluses I don't think? Can we use commas or something?

Also surprised that we're using display names, I thought these were usually short like three-letter codes?

Of course, anything we do to change URLs needs to support all legacy URLs as much as possible. URLs are an API you can never deprecate…

**Comment by danielgavrilov** [comment]
> We shouldn't be using pluses I don't think? Can we use commas or something?

Entity names can contain commas... 🙂 But maybe `|` or something else could work, or we could escape commas within entity names.

> [...] I thought these were usually short like three-letter codes?

Yes they are, but we fall back to the names when those are not available (e.g. for continents).

> Of course, anything we do to change URLs needs to support all legacy URLs as much as possible. URLs are an API you can never deprecate…

Definitely the big issue! 

---

Two simple approaches that might work here:

1. Use underscores `_` instead of `%20` for spaces. We can be backwards-compatible by interpreting strings in multiple ways, e.g. `South_America` becomes `["South_America", "South America"]` and we discard any invalid entity names.
2. Use `|` for separators instead of `+`. We then split on both `|` and `+` to get a list from a string. We currently have no entities using `|` in their name, but we could also implement logic to escape it, just to be safe.

Another related thing I noticed, we have 102 entities in the database that contain a `+` – here are some of them:

```
Men and Women Ages 65+
Men and Women Ages 18+
Western Africa + (Total)
Western Asia + (Total)
Western Europe + (Total)
World + (Total)
Africa + (Total)
Americas + (Total)
Asia + (Total)
Australia & New Zealand + (Total)
Caribbean + (Total)
Central America + (Total)
Central Asia + (Total)
...
```

So we already have an issue with using `+` to separate entity names in URLs, unless we escape them (I don't think we do?)

**Comment by breck7** [comment]
Adding another repro. We got another report of this for the link "https://ourworldindata.org/grapher/covid-confirmed-deaths-since-5th-death?country=CAN+USA"

**Comment by jimsmart** [comment]
I've encountered this, and can reproduce it all the time — I came here to file an issue myself. It's a particularly annoying bug!

I don't think you need to change the link format (which would break existing/legacy links), but instead I think/believe the issue really lies where/how the links are decoded. 

Is that something that your code is handling? Client-side or server-side? Or is this in some third-part library. Happy to take a look myself, if someone can point me in the right direction.

To me, a brief look seems to indicate that the country params are being handled server-side(?), and being somehow unparsable in their incoming state are then set to "", and then a redirect is performed. (I've not confirmed this at the request-response to see this, I am merely inferring it from the final resulting URL having `country=` at the end of it)

Ideally, the URL decoder should be made to handle both + and %20, they should both decode to an ASCII space.

— Great work on all the whole project, its data and its presentation! I look forward to being able to share the links more freely / with less bother!

**Comment by jimsmart** [comment]
FWIW, both + and %20 are valid encodings according to RFC3986, what FB are doing here isn't against the spec, even it is somewhat unexpected.

**Comment by jimsmart** [comment]
[In this function](https://github.com/owid/owid-grapher/blob/192305fb33b2ee568d6fecd1d21265325e48f815/utils/client/url.ts#L70), to decode the query params, a simple regex replacement is performed.

I think the fix is to use the built-in [decodeURI()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURI) method on the original incoming URL, before the query string is stripped off, at a higher level than this method.

[This function](https://github.com/owid/owid-grapher/blob/02544581116cda4d6047f8c790f9e963f2d6e522/charts/ChartUrl.ts#L335) may also need attention? If URL decoding is done at a higher level than this function, it might then be dealing with spaces here instead of + here.

**Comment by jimsmart** [comment]
Ah. I now realise that I was only thinking about country codes, and not country names with spaces and pluses. :neutral_face:

**Comment by breck7** [comment]
Thanks for investigating @jimsmart !

I've started a branch to create a solution for this (https://github.com/owid/owid-grapher/tree/country-urls)

**Comment by breck7** [comment]
Here's another test case:

This link:
https://ourworldindata.org/grapher/share-of-violent-deaths-prehistoric-archeological-sites?country=Bogebakken%20(Denmark)%3B%204300%20-%203800%20BCE+British%20Columbia%20(30%20sites)%3B%203500%20BCE%20-%201674%20CE+Brittany%3B%206000%20BCE

Becomes this on Facebook:
https://ourworldindata.org/grapher/share-of-violent-deaths-prehistoric-archeological-sites?country=Bogebakken+%28Denmark%29%3B+4300+-+3800+BCE+Brittany%3B+6000+BCE+British+Columbia+%2830+sites%29%3B+3500+BCE+-+1674+CE

**Comment by breck7** [comment]
Another test case.

Facebook turns this:
https://playfair-owid.netlify.app/grapher/share-population-female?tab=chart&country=Caribbean%20small%20states

Into this:
https://playfair-owid.netlify.app/grapher/share-population-female?tab=chart&country=Caribbean+small+states

**Comment by MarcelGerber** [comment]
Should be fixed now with #446.
